File: app/src/main/java/net/numa08/gochisou/data/model/PageNation.java
package net.numa08.gochisou.data.model;

import com.google.gson.annotations.SerializedName;

import java.util.List;

public abstract class PageNation<T> {
    private List<T> list;
    @SerializedName("prev_page")
    private int prevPage;
    @SerializedName("next_page")
    private int nextPage;
    @SerializedName("total_count")
    private int totalCount;

    public List<T> getList() {
        return list;
    }

    public int getPrevPage() {
        return prevPage;
    }

    public int getNextPage() {
        return nextPage;
    }

    public int getTotalCount() {
        return totalCount;
    }

    public static class TeamPageNation extends PageNation<Team> {
        @SerializedName("teams")
        private List<Team> teams;

        @Override
        public List<Team> getList() {
            return teams;
        }
    }
}

File: app/src/main/java/net/numa08/gochisou/data/model/PageNation.kt
package net.numa08.gochisou.data.model

import com.google.gson.annotations.SerializedName

sealed class PageNation<T> {
    open val list: List<T>? = null
    @SerializedName("prev_page")
    val prevPage: Int = 0
    @SerializedName("next_page")
    val nextPage: Int = 0
    @SerializedName("total_count")
    val totalCount: Int = 0

    class TeamPageNation : PageNation<Team>() {
        @SerializedName("teams")
        override val list: List<Team>? = null
    }

    class PostPageNation : PageNation<Post>() {
        @SerializedName("posts")
        override val list: List<Post>? = null
    }

    class MemberPageNation : PageNation<Member>() {
        @SerializedName("members")
        override val list: List<Member>? = null
    }
}
